### PR TITLE
Add Activate Service Package Empty View

### LINKED ISSUE #4458: Display Empty View with some information if there's no available service packages to activate.
This is a specific use case to support PR #4433

### PR DESCRIPTION
**Pull request by hafizrahman**
Fixes #4458 

# What this PR does:

#4433 adds the feature to activate a service package (also known as predefined package), during shipping label creation process.

In the case where a user has activated all available service packages, then the activate service package screen should show an Empty View informing the user that there's nothing to activate anymore.

# Testing Instruction:

**Regular case test:**

0. Make sure the test site has Shipping Labels creation enabled, and have an Order where Shipping Labels can be created.
1. Activate all service packages in wp-admin: ` wp-admin/admin.php?page=wc-settings&tab=shipping&section=woocommerce-services-setting`
2. In the app, find the Order from step 0 and create a Shipping Label, complete the "Ship from" and "Ship to" steps, then continue with the "Packaging details" step.
3. Tap the "Package selected" spinner, this will open the "Package selection" screen. Tap "Add New Package" at the bottom.
4. On "Package creation" screen, switch to the "Service Package" tab. This will show a loading screen, and then finally show the empty view. Make sure the "Done" menu item is not shown at the top.

**Cache test:**
0. Make sure the test site has Shipping Labels creation enabled, and have an Order where Shipping Labels can be created.
1. Activate all service packages in wp-admin: ` wp-admin/admin.php?page=wc-settings&tab=shipping&section=woocommerce-services-setting`, but keep one service package inactivated
2. In the app, find the Order from step 0 and create a Shipping Label, complete the "Ship from" and "Ship to" steps, then continue with the "Packaging details" step.
3. Tap the "Package selected" spinner, this will open the "Add new package" screen.
4. Switch to the "Service Package" tab. Select the one remaining package, then tap Done to activate it.
5. You will be navigated back to the package editing screen.
6. Tap the "Package selected" spinner, this will open the "Package selection" screen. Tap "Add New Package" at the bottom.
7. On "Package creation" screen, switch to the "Service Package" tab. This will show the empty view. Make sure the "Done" menu item is not shown at the top.

| Light mode | Dark mode |
|-|-|
| image | image |

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/ShippingLabelCreateServicePackageFragment.kt [hunk 1]
```diff
@@ -63,6 +70,20 @@ class ShippingLabelCreateServicePackageFragment :
...
+            new.isEmpty.takeIfNotEqualTo(old?.isEmpty) { isEmpty ->
+                if (isEmpty) {
+                    binding.errorView.show(
+                        type = WCEmptyView.EmptyViewType.SHIPPING_LABEL_SERVICE_PACKAGE_LIST
+                    )
+                } else {
+                    binding.errorView.hide()
+                }
+
+                binding.servicePackagesListContainer.isVisible = !isEmpty
```
Comment (malinajirka) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/ShippingLabelCreateServicePackageFragment.kt]: ⚠️ The business logic related to the "Done" button is currently being handled in the view layer. I believe it's a good practice to keep the view layer as simple as possible -> we might want to consider moving this logic to the vm layer and write tests (aka documentation) for it. (Notice, that we currently can't write unit tests for the visibility of the "Done" button since the logic is in the, almost un-testable, view layer).
Perhaps something like [this](https://github.com/woocommerce/woocommerce-android/blob/e19a9c0b1d810135e4f094557e1302c9164e5d55/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/EditShippingLabelPaymentFragment.kt#L170) might work, wdyt?

**Review by malinajirka — CHANGES_REQUESTED** [review]
Thanks for working on this @hafizrahman and for providing great testing instructions even for someone who doesn't have much experience with shipping labels.

I noticed one issue where I end up on a blank screen:
1. I followed both the test scenarios in the test description
2. I cleared app's cache and followed the "Cache test:" - I expected to see the non-activated service package (since I cleared the cache), but I ended up on a blank screen instead


I also left one comment about a business logic in the view layer. I realize the VM doesn't contain any unit tests yet, but I'd consider adding them - at least for the logic you touched in this PR. I believe we shouldn't introduce new code without unit tests as it's very difficult to maintain and very prone to re-introducing bugs. It is also a great way how to document the code.

P.S. I'm wondering if we should introduce a "retry" button which would skip the cache on the empty screen? Also it's not related to this PR, but it seems the in-memory cache is indefinite -> we never clear it unless the app crashes or the system/user kills the app -> I've seen apps running for months in the background, so I'm wondering if it's safe to use an indefinite in-memory-cache. Wdyt?

**Comment by hafizrahman** [comment]
@malinajirka Thank you for testing and sharing your feedback!

With the screenshot that you shared, was that on step 4 of the Cache test (when the app should've shown 1 package to active), or step 7 (when the app should've shown the empty view)?

And to clarify, did you do "Regular case test", then clear app cache, then do "Cache test"?

Also in case I made things confusing by calling the second test phase "Cache step", I want to clarify that the "caching" part was referring to this [caching that's happening in the repository](https://github.com/woocommerce/woocommerce-android/blob/6918bbd73876dc0d87aeeeab4abbd19cd11eedf2/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/ShippingLabelRepository.kt#L91) so the app doesn't call the API each time the screen is loaded.

> it seems the in-memory cache is indefinite -> we never clear it unless the app crashes or the system/user kills the app -> I've seen apps running for months in the background, so I'm wondering if it's safe to use an indefinite in-memory-cache. Wdyt?'

Thanks for mentioning that, that is also something that I had in mind. I noticed that the current state might not be ideal in the (rare but possible) case of where user is editing the packages in wp-admin while working with the app at the same time; the wp-admin changes won't be reflected in the app. I'll bring it up for conversations with Hicham and see how we can better handle that. 

As for unit tests, I agree with you and for that I've [set a separate issue](https://github.com/woocommerce/woocommerce-android/issues/4464) and plan to tackle everything at once next week. My initial concern was that the PRs related to the whole activate/create package thing were too big, so I split the unit tests away.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/ShippingLabelCreateServicePackageFragment.kt (hunk 1 above)
Comment (hafizrahman) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/ShippingLabelCreateServicePackageFragment.kt]: I thought I already did this by putting the logic for both showing the empty view and hiding the menu item based on the state of `isEmpty` 🤔 

I looked at your example and it seems to be implementing the same idea, can you clarify a bit more on this?

Comment (malinajirka) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/ShippingLabelCreateServicePackageFragment.kt]: I see your point ;)! You are probably right - my comment was too nitpicky, so feel free to ignore it :).

Having said that I feel like I should clarify what I was thinking. Notice how is the "canSave" field implemented in the example I shared. 
```
val canSave: Boolean
            get() = canEditSettings && paymentMethods.any { it.isSelected }
```
so I was thinking we could add a field into the viewState, something like
```
val canSave: Boolean
            get() = !isEmpty
```.
The fact that we want to hide the done button when the list is empty is a bit hidden in the current implementation -> when you look at the ViewState you don't see that there. 
I mostly like to follow 1:1 mapping as if some of the fields in the viewstate are not used in the view layer -> that's an obvious bug. However, when you map some of the fields (such as "isEmpty") to multiple states (such as the empty view and the done button) then noticing that we forgot something in the view layer isn't that obvious anymore.

Hope this clarification makes sense, but as I said, feel free to keep the code as is.

**Comment by malinajirka** [comment]
> With the screenshot that you shared, was that on step 4 of the Cache test (when the app should've shown 1 package to active), or step 7 (when the app should've shown the empty view)?

I'm a complete mess, sorry 😞  -> I must have switched from jalapeno app to debug app so when I saw the blank screen issue, I wasn't using the version of the app from this PR. I'm truly sorry for wasting your time.

> I noticed that the current state might not be ideal in the (rare but possible) case of where user is editing the packages in wp-admin while working with the app at the same time;

I don't think they need to be using wpadmin and the app at the same time. I think they can 
1. run the app, add a package service
2. week later edit wpadmin
3. week later re-open the app (the OS might keep it open the whole time since step 1)
4. They'll see the state from step 1 even though they changed the settings last week

It is still an edge case scenario, but I'd consider setting validity of the cache to something like 2 hours. Either case, definitely not in the scope of this PR ;).


> As for unit tests, I agree with you and for that I've set a separate issue and plan to tackle everything at once next week. My initial concern was that the PRs related to the whole activate/create package thing were too big, so I split the unit tests away.

Awesome, thank you 🙇 !!! 
FWIW when it comes to the size of the PR - I personally find that tests help the reviewer trust the code more and serve as a documentation even during the review so I wouldn't worry about the size of the PR when the size is big because of the tests. If you feel that'd be confusing, I'd consider creating a separate PR for the tests at the same time as you open the main PR - so it's up to the reviewer if they want to read the tests first or not.


> 4. Switch to the "Service Package" tab. Select the one remaining package, then tap Done to activate it.

Is the step `4` in the cache test really feasible? I don't see the "one remaining package" since the app used the cache -> I see the  "empty state". The app doesn't know that I kept `one service package inactivated` in step `1` in WPAdrmin. Am I missing something?

**Review by malinajirka — APPROVED** [review]
I think this works as intended - I think there is just some confusion in the testing steps. I'm approving this PR and feel free to merge it into 7.2. Great job working on this 👏 !

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/ShippingLabelCreateServicePackageFragment.kt (hunk 1 above)
Comment (hafizrahman) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/ShippingLabelCreateServicePackageFragment.kt]: > The fact that we want to hide the done button when the list is empty is a bit hidden

Ah I see! I like that line of thinking and I agree. Will make the changes.

**Comment by hafizrahman** [comment]
> Is the step 4 in the cache test really feasible? I don't see the "one remaining package" since the app used the cache -> I see the "empty state". The app doesn't know that I kept one service package inactivated in step 1 in WPAdrmin. Am I missing something?

> I think there is just some confusion in the testing steps. 

Yup I think these two are related! I didn't intend for the two test cases to be run back-to-back. In that situation the one remaining package set in wp-admin on case 2 won't be shown in the package list, b/c it's already caching the "no existing package" state from case 1. That's probably what you're seeing and mentioning, so sorry for the confusion!

> If you feel that'd be confusing, I'd consider creating a separate PR for the tests at the same time as you open the main PR - so it's up to the reviewer if they want to read the tests first or not.

Nice idea, thanks for sharing that! I'll try that in the future.

**Review by malinajirka — APPROVED** [review]
> Yup I think these two are related! I didn't intend for the two test cases to be run back-to-back. In that situation the one remaining package set in wp-admin on case 2 won't be shown in the package list, b/c it's already caching the "no existing package" state from case 1. That's probably what you're seeing and mentioning, so sorry for the confusion!

Ahh I see, makes sense! Thanks for the clarification!

**Comment by hichamboushaba** [comment]
Thanks @hafizrahman for working on this, and thanks @malinajirka for reviewing it 🙏 

> Also it's not related to this PR, but it seems the in-memory cache is indefinite -> we never clear it unless the app crashes or the system/user kills the app

I didn't check all the code, nor followed the whole discussion, so I may be wrong here, but unless you are talking about something else, I don't think we keep the cache indefinitely, we clear it once the user leaves the shipping labels form: https://github.com/woocommerce/woocommerce-android/blob/adab8e8a5905ed6132666b3f5d95805b0f922766/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/ShippingLabelRepository.kt#L268-L271

So unless they start the process of creating a label, then they leave it open for long time before coming back to continue, it won't cause any issues.

**Comment by malinajirka** [comment]
I totally missed that, good point @hichamboushaba!
